src/server/db_slice.cc

```diff
@@ -1096,10 +1101,70 @@ auto DbSlice::DeleteExpiredStep(const Context& cntx, unsigned count) -> DeleteEx
   return result;
 }
 
-// TODO: Design a better background evicting heuristic.
+int32_t DbSlice::GetNextSegmentForEviction(int32_t segment_id, DbIndex db_ind) const {
+  // wraps around if we reached the end
+  return db_arr_[db_ind]->prime.NextSeg((size_t)segment_id) %
+         db_arr_[db_ind]->prime.GetSegmentCount();
+}
+
 void DbSlice::FreeMemWithEvictionStep(DbIndex db_ind, size_t increase_goal_bytes) {
   if (!caching_mode_)
     return;
+
+  auto time_start = absl::GetCurrentTimeNanos();
+  auto& db_table = db_arr_[db_ind];
+  int32_t num_segments = db_table->prime.GetSegmentCount();
+  int32_t num_buckets = PrimeTable::Segment_t::kTotalBuckets;
+  int32_t num_slots = PrimeTable::Segment_t::kNumSlots;
+
+  size_t used_memory_after;
+  size_t evicted = 0;
+
+  size_t used_memory_before = owner_->UsedMemory();
+  for (int32_t slot_id = num_slots - 1; slot_id >= 0; --slot_id) {
+    for (int32_t bucket_id = num_buckets - 1; bucket_id >= 0; --bucket_id) {
+      // pick a random segment to start with in each eviction,
+      // as segment_id does not imply any recency, and random selection should be fair enough
+      int32_t segment_id = rand() % num_segments;
+      for (size_t num_seg_visited = 0; num_seg_visited < max_segment_to_consider_;
+           ++num_seg_visited, segment_id = GetNextSegmentForEviction(segment_id, db_ind)) {
+        const auto& bucket = db_table->prime.GetSegment(segment_id)->GetBucket(bucket_id);
```

Comment:
> just a few large keys as it will take lots of iterations to find them

Yes, that's a great point, my testing only uses strings, and I did not really test such situation. I do plan to test things like hash set where a single key could introduce a large amount of memory usage. I will try to include such testing cases in the python tool later.

> I assume there are small optimizations like remembering a few non-empty buckets when looking up a different slot inside them just to evict there if we didn't find anything else

Yes, actually I have a question, do you know if we are maintaining the metadata such as emptiness of the buckets? If so, I certainly can leverage those information easily.

Reply:
I'm not an expert with dashtable code, but I assume that we don't have this info right at hand. What I referred to was getting a bucket's GetBusy() bitmask to determine quickly what slots really are occupied (instead of pointing at just one)

But you should really first benchmark it with a single key before changing the implementation. I just did the estimation of 3k lookups with a default heartbeat frequency of 100/s which adds up to 300k ops/second which is... a lot 😆